### ISSUE TITLE
Minio cache not working on single file chunked multithreaded downloads via rclone

### ISSUE THREAD

**Issue by dpgarrick** [minio/minio#9597]
I set up a Minio s3 gateway with disk cache. Everything functioned as expected except when trying to download files via rclone >250MB. Searching further, rclone by default downloads files > 250 MB in a multithreaded manner with each thread downloading some chunk of the file.

For such downloads the cache was being populated with multiple chunks which presumably matched up to those that rclone had requested. However, any subsequent download of the same file resulted in those cache files being immediately cleared and redownloaded again, and the cached files never being served even when they exist.

However, if a cache was populated from a download with rclone "--multi-thread-stream 0" as a single thread download, the cache was in that case properly populated with the whole file. Any subsequent downloads, including multi-threaded downloads, were in that case then always served from the cache successfully.
## Expected Behavior

Minio should cache and serve cached files the same way regardless of whether the download which initiated the caching was multithreaded (chunked) or not.

## Current Behavior

Minio not serving (not recognising?) file is properly cached when the download which triggered the cache was initiated as a multi-threaded download, which is the default behaviour for "rclone copy" for files > 250 MB. 

Cache of such files is repeatedly cleared and redownloaded without being served when subsequent downloads of the same file are initiated.

## Possible Solution

Current workaround is to disable rclone multithread streams using the flag "--multi-thread-stream 0"

If such download triggers the cache, subsequent downloads are successfully served from the cache even if --multi-thread-stream > 0 but since there is no easy to way know if a file is cached, it just needs to be set to 0 always.

If I had to guess, the issue is related to a checksum check on the cached files. When the cache is triggered from a multithreaded rclone copy, inspecting the cache shows files of equal chunks to the chunks requested by rclone. Obviously the checksum of those chunks would not match the original file so I would guess the cache is invalidated by Minio on subsequent downloads as it assumes the file has changed at source s3?

If the cache holds the whole file, then Minio does serve it correctly even if the subsequent downloads are multithreaded. So seems to me a solution sould be if Minio knew to cache the whole file even when the download that triggered it was a chunked multithreaded download, then the problem would be fixed

## Steps to Reproduce (for bugs)

1. Define cache settings for an s3 gateway
```
MINIO_CACHE="on"
MINIO_CACHE_DRIVES="/minio/cache"
MINIO_CACHE_QUOTA=80
MINIO_CACHE_AFTER=0
MINIO_CACHE_WATERMARK_LOW=70
MINIO_CACHE_WATERMARK_HIGH=90
MINIO_ACCESS_KEY=access_key
MINIO_SECRET_KEY=secret_key
```
2. Run Minio
```
minio gateway s3
```
2.  Download rclone from rclone.org, set up rclone.conf to gateway
```
[ms3]
type = s3
provider = Minio
env_auth = false
access_key_id = access_key
secret_access_key = secret_key
endpoint = localhost:9000
```
3. On a separate terminal set up a "cache watcher" to view the changes in the cache:
```
watch  --color --differences --interval .1 --no-title du -ah /minio/cache/
```
3. Multithreaded download  - first create a 20MB file.bin on the s3, then
```
rclone --config rclone.conf copy ms3:file.bin . -vvv --multi-thread-cutoff 2M --multi-thread-streams 4
```
4. Check cache watcher window. Should see 4 equally split ~5MB chunks download to cache as the file downloads (my actual file for testing is 17MB)
```
4.0K	/minio/cache/.minio.sys/format.json
8.0K	/minio/cache/.minio.sys
4.1M	/minio/cache/ec45a1ac4856772aa6a824bc21e5d2288ea8d0879b392afa4886ab816e333632/1085fc64-b847-4242-8f4c-633e37853247
4.2M	/minio/cache/ec45a1ac4856772aa6a824bc21e5d2288ea8d0879b392afa4886ab816e333632/c8954e6d-3a39-4622-aafb-8516c9d134ed
4.0K	/minio/cache/ec45a1ac4856772aa6a824bc21e5d2288ea8d0879b392afa4886ab816e333632/cache.json
4.2M	/minio/cache/ec45a1ac4856772aa6a824bc21e5d2288ea8d0879b392afa4886ab816e333632/208d806e-e636-4f11-91aa-4067c114791c
4.2M	/minio/cache/ec45a1ac4856772aa6a824bc21e5d2288ea8d0879b392afa4886ab816e333632/ef89d05a-28e5-49aa-b8aa-90b7dff0cda5
17M	/minio/cache/ec45a1ac4856772aa6a824bc21e5d2288ea8d0879b392afa4886ab816e333632
17M	/minio/cache
```
5. Remove local file and download again. Check the cache watcher window - all the chunks in the cache first disappear and are then redownloaded as the file downloads.
```
rm ./file.bin
rclone copy --config rclone.conf ms3:file.bin . -vvv --multi-thread-cutoff 2M --multi-thread-streams 4
```

6. Remove local file and download again in single thread. Check the cache watcher window - a full size file will download into the cache and then the four chunks will disappear on completion.
```
rm ./file.bin
rclone copy --config rclone.conf ms3:file.bin . -vvv --multi-thread-cutoff 2M --multi-thread-streams 1
```
```
4.0K	/minio/cache/.minio.sys/format.json
8.0K	/minio/cache/.minio.sys
17M	/minio/cache/ec45a1ac4856772aa6a824bc21e5d2288ea8d0879b392afa4886ab816e333632/part.1
4.0K	/minio/cache/ec45a1ac4856772aa6a824bc21e5d2288ea8d0879b392afa4886ab816e333632/cache.json
17M	/minio/cache/ec45a1ac4856772aa6a824bc21e5d2288ea8d0879b392afa4886ab816e333632
17M	/minio/cache
```
7. Remove local file and download with/without multiple chunks/threads. No changes in cache - cached file is now being served to all downloads regardless of whether they are chunked or not.
```
rm ./file.bin
rclone copy --config rclone.conf ms3:file.bin . -vvv --multi-thread-cutoff 2M --multi-thread-streams 4
```

## Context

I have a local analytics cluster which downloads and uploads various data files from an AWS s3 bucket. Some of the files generated and uploaded from one part of the pipeline are downloaded by another stage for other analysis.

I set one of these local machines up as a minio s3 gateway so that data is locally cached rather than having each server individually downloading all files directly from the s3 bucket which is slower and adds up transfer costs.

For files > 250MB I wasn't getting the download speeds I was expecting to see from a locally cached file until I realised rclone was doing chunked downloads by default for files >250MB and discovered the caching issue. Setting rclone --multi-thread-streams 0 "fixes" the problem so not a critical issue but default rclone behaviour is to do this chunked download so anyone else using rclone with minio cache would not be getting the benefits of the disk cache. 

Potentially if there are other s3 tools that perform multithreaded chunked downloads the cache may also be failing in those cases too. Not sure if it also occurs on chunked uploads.

## Regression

Unknown

## Your Environment

* Version used (`minio version`):
`minio version RELEASE.2020-05-08T02-40-49Z`
```
rclone v1.51.0
- os/arch: linux/amd64
- go version: go1.13.7
```
* Operating System and version (`uname -a`):
`5.3.0-51-generic #44~18.04.2-Ubuntu SMP Thu Apr 23 14:27:18 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by poornas** [comment]
disk cache implementation does not concatenate chunks back into the original object if that is what you are looking for. The cache implementation either stores range GET chunks or the entire object (in which case the chunks are deleted) as it avoids overhead of managing overlapping ranges and provides a consistent and simple approach.

**Comment by dpgarrick** [comment]
I see that the range GET chunks are cached, the issue is that if I do subsequent downloads using an identical rclone copy command which as far as I can tell uses identical range GET chunks as the one that was cached, the cache is not serving those chunks but clearing them out and repeatedly redownloading them on each download

**Comment by harshavardhana** [comment]
> I see that the range GET chunks are cached, the issue is that if I do subsequent downloads using an identical rclone copy command which as far as I can tell uses identical range GET chunks as the one that was cached, the cache is not serving those chunks but clearing them out and repeatedly redownloading them on each download

@dpgarrick how do you confirm that it's using the same range? how much disk space do you have locally? do you have any cache-control headers set for these objects?

**Comment by dpgarrick** [comment]
@harshavardhana rclone -vv (verbose) output for subsequent downloads lists the ranges as being the same.

```
2020-05-14 15:26:05 DEBUG : file.csv.bz2: Need to transfer - File not found at Destination
2020-05-14 15:26:05 DEBUG : file.csv.bz2: Starting multi-thread copy with 4 parts of size 4.188M
2020-05-14 15:26:05 DEBUG : file.csv.bz2: multi-thread copy: stream 4/4 (13172736-17441221) size 4.071M starting
2020-05-14 15:26:05 DEBUG : file.csv.bz2: multi-thread copy: stream 2/4 (4390912-8781824) size 4.188M starting
2020-05-14 15:26:05 DEBUG : file.csv.bz2: multi-thread copy: stream 1/4 (0-4390912) size 4.188M starting
2020-05-14 15:26:05 DEBUG : file.csv.bz2: multi-thread copy: stream 3/4 (8781824-13172736) size 4.188M starting
2020-05-14 15:26:06 DEBUG : file.csv.bz2: multi-thread copy: stream 3/4 (8781824-13172736) size 4.188M finished
2020-05-14 15:26:06 DEBUG : file.csv.bz2: multi-thread copy: stream 1/4 (0-4390912) size 4.188M finished
2020-05-14 15:26:06 DEBUG : file.csv.bz2: multi-thread copy: stream 2/4 (4390912-8781824) size 4.188M finished
2020-05-14 15:26:07 DEBUG : file.csv.bz2: multi-thread copy: stream 4/4 (13172736-17441221) size 4.071M finished
2020-05-14 15:26:07 DEBUG : file.csv.bz2: Finished multi-thread copy with 4 parts of size 4.188M
2020-05-14 15:26:07 DEBUG : file.csv.bz2: MD5 = 5d9e48adcbf2be4bb8ed3b7dbc599dae OK




2020-05-14 15:27:40 DEBUG : file.csv.bz2: Need to transfer - File not found at Destination
2020-05-14 15:27:40 DEBUG : file.csv.bz2: Starting multi-thread copy with 4 parts of size 4.188M
2020-05-14 15:27:40 DEBUG : file.csv.bz2: multi-thread copy: stream 4/4 (13172736-17441221) size 4.071M starting
2020-05-14 15:27:40 DEBUG : file.csv.bz2: multi-thread copy: stream 2/4 (4390912-8781824) size 4.188M starting
2020-05-14 15:27:40 DEBUG : file.csv.bz2: multi-thread copy: stream 1/4 (0-4390912) size 4.188M starting
2020-05-14 15:27:40 DEBUG : file.csv.bz2: multi-thread copy: stream 3/4 (8781824-13172736) size 4.188M starting
2020-05-14 15:27:41 DEBUG : file.csv.bz2: multi-thread copy: stream 4/4 (13172736-17441221) size 4.071M finished
2020-05-14 15:27:41 DEBUG : file.csv.bz2: multi-thread copy: stream 3/4 (8781824-13172736) size 4.188M finished
2020-05-14 15:27:41 DEBUG : file.csv.bz2: multi-thread copy: stream 2/4 (4390912-8781824) size 4.188M finished
2020-05-14 15:27:41 DEBUG : file.csv.bz2: multi-thread copy: stream 1/4 (0-4390912) size 4.188M finished
2020-05-14 15:27:41 DEBUG : file.csv.bz2: Finished multi-thread copy with 4 parts of size 4.188M
2020-05-14 15:27:41 DEBUG : file.csv.bz2: MD5 = 5d9e48adcbf2be4bb8ed3b7dbc599dae OK
2020-05-14 15:27:41 INFO  : file.csv.bz2: Multi-thread Copied (new)
```

The drive the cache is on is only 7% full (500GB SSD), well below the minio cache settings for clearing out. No cache control headers set that I am aware of.

**Comment by dpgarrick** [comment]
What I am looking for is consistency - currently if my download which triggers the caching is done as a whole object the cache saves the whole object.

Subsequent downloads, even if chunked, are served from that cached whole object without issue and that cached whole object never disappears no matter how many different ways I download that file, chunked or unchunked.


But if the initial download which triggers the cache is a chunked download, the cache generates chunked ranges instead of the whole object. I don’t have any problem with that necessarily but the behaviour is inconsistent - if I perform an identical download against the same chunks, those chunks in the cache are flushed (whether they are served for the download or not first I do not know) and then subsequently freshly recached.

If I store the freshly recached chunks between subsequent identical chunked downloads they are identical:

```
dpg@h:~$ md5sum cache/*
60083b154d924f4301f9b9f6e70b350f  cache/2f5cad54-8d10-44a7-8aec-4d78350e7f5b
116ceccaea529cecb8b7be719d184976  cache/3c4a9095-97b2-400b-b98f-86c7918957e6
659f7e446f2b89933b8ad7e83a77f3f9  cache/5e003a81-5edb-45bb-a821-7b329dff55c2
4dc3e0e036dcdb8b3dbb3ee776b5b636  cache/5f3be6de-1202-401c-9c77-cd5d57d49eb5
8cd24a97a1868f46549057025adec167  cache/cache.json
```
```
dpg@h:~$ md5sum cache2/*
60083b154d924f4301f9b9f6e70b350f  cache2/237d379e-1175-4196-a7e5-504cb930095b
116ceccaea529cecb8b7be719d184976  cache2/6a59d45c-8037-4f5a-bfc6-e3fe6bb7247c
4dc3e0e036dcdb8b3dbb3ee776b5b636  cache2/7de97a98-1c0f-4c4a-b1ad-32d5455f7088
659f7e446f2b89933b8ad7e83a77f3f9  cache2/9d27db9d-d3d4-4177-858f-56b9c83a67e0
6775ada6b5e0086c3cde9c5f68960051  cache2/cache.json
```

Maybe there is some weird interaction between rclone's ranged GETs and minio but otherwise what I am looking for would either be that subsequent identically ranged GET downloads (via rclone copy) are properly served from the cache or some kind of additional CACHE environment variable I can set so that ranged GET requests on an object cause the whole object to be stored in cache instead of the equivalent ranged GET chunks since that seems to be broken while subsequent ranged GET downloads on the whole object are successfully served from cache.

**Comment by dpgarrick** [comment]
The cache.json files on the two caches if it helps...

```
dpg@h:~$ cat cache/cache.json | jq
{
  "version": "1.0.0",
  "stat": {
    "size": 17441221,
    "modTime": "2020-05-14T03:26:08.726672554Z"
  },
  "checksum": {
    "algorithm": "highwayhash256S",
    "blocksize": 1048576
  },
  "meta": {
    "Content-Type": "application/octet-stream",
    "X-Amz-Meta-Mtime": "1588539339",
    "content-type": "application/octet-stream",
    "etag": "5d9e48adcbf2be4bb8ed3b7dbc599dae",
    "expires": "Mon, 01 Jan 0001 00:00:00 GMT"
  },
  "ranges": {
    "0-4390911": "2f5cad54-8d10-44a7-8aec-4d78350e7f5b",
    "13172736-17441220": "5e003a81-5edb-45bb-a821-7b329dff55c2",
    "4390912-8781823": "3c4a9095-97b2-400b-b98f-86c7918957e6",
    "8781824-13172735": "5f3be6de-1202-401c-9c77-cd5d57d49eb5"
  },
  "hits": 4
}
```

```
dpg@h:~$ cat cache2/cache.json | jq
{
  "version": "1.0.0",
  "stat": {
    "size": 17441221,
    "modTime": "2020-05-14T03:38:19.721836338Z"
  },
  "checksum": {
    "algorithm": "highwayhash256S",
    "blocksize": 1048576
  },
  "meta": {
    "Content-Type": "application/octet-stream",
    "X-Amz-Meta-Mtime": "1588539339",
    "content-type": "application/octet-stream",
    "etag": "5d9e48adcbf2be4bb8ed3b7dbc599dae",
    "expires": "Mon, 01 Jan 0001 00:00:00 GMT"
  },
  "ranges": {
    "0-4390911": "237d379e-1175-4196-a7e5-504cb930095b",
    "13172736-17441220": "9d27db9d-d3d4-4177-858f-56b9c83a67e0",
    "4390912-8781823": "6a59d45c-8037-4f5a-bfc6-e3fe6bb7247c",
    "8781824-13172735": "7de97a98-1c0f-4c4a-b1ad-32d5455f7088"
  },
  "hits": 12
}
```

**Comment by dpgarrick** [comment]
Started playing around with a bigger file (500MB) as follows
1. Download file using rclone default copy settings (ranged GET). Verified ranged chunks exist in cache.
2. Redownload with identical command (identical ranged GET). Cached chunks disappear as soon as I start the download and cache is now empty
3. Redownload with identical command again (ranged GET). Verified new ranged chunks exist in cache. Different IDs but identical ranges and md5sums as the previous cached chunks.
4. Return to step 2 as this cycle repeats....

If at any time I download the full object with --multi-thread-stream 1 the whole object is cached once and all future requests seem to be served from cache regardless of whether subsequent downloads are ranged GETs or not.

It is not really critical for me since I have just set all my rclone scripts to use --multi-thread-stream 1 now due to this issue but it definitely seems the ranged GET cache functionality has a bug. 

Happy to try things to help debug, just let me know.
